spec/unit/controllers/v3/service_instance_controller_spec.rb

```diff
@@ -0,0 +1,198 @@
+
+        expect(response.status).to eq(403)
+        expect(response.body).to include('FeatureDisabled')
+        expect(response.body).to include('service_instance_sharing')
+      end
+    end
+
+    context 'when the service instance does not exist' do
+      it 'returns a 404' do
+        post :share_service_instance, service_instance_guid: 'nonexistant-service-instance-guid', body: req_body
+        expect(response.status).to eq 404
+        expect(response.body).to include('Service instance not found')
+      end
+    end
+
+    context 'when the target space does not exist' do
+      before do
+        req_body[:data] = [{ guid: 'nonexistant-space-guid' }]
+      end
+
+      it 'returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+        expect(response.body).to include('Unable to share to spaces')
+        expect(response.body).to include('nonexistant-space-guid')
+      end
+    end
+
+    context 'when multiple target spaces do not exist' do
+      before do
+        req_body[:data] = [
+          { guid: 'nonexistant-space-guid' },
+          { guid: 'nonexistant-space-guid2' },
+          { guid: target_space.guid }
+        ]
+      end
+
+      it 'does not share to any of the valid spaces and returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+        expect(response.body).to include('Unable to share to spaces')
+        expect(response.body).to include('nonexistant-space-guid')
+        expect(response.body).to include('nonexistant-space-guid2')
+        expect(service_instance.shared_spaces).to_not include(target_space)
+      end
+    end
+
+    context 'when the service instance has already been shared with the specified space' do
+      before do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+      end
+
+      it 'returns a 200 and leaves the existing share intact' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+
+        expect(response.status).to eq 200
+        expect(service_instance.shared_spaces).to include(target_space)
+      end
+    end
+
+    context 'when the request is malformed' do
+      let(:req_body) {
+        {
+          bork: 'some-name',
+        }
+      }
+
+      it 'returns a 422' do
+        post :share_service_instance, service_instance_guid: service_instance.guid, body: req_body
+        expect(response.status).to eq 422
+      end
+    end
+
+    describe 'permissions by role' do
+      context 'when the user is a space developer in the source space' do
+        before do
+          set_current_user_as_role(role: 'space_developer', org: source_space.organization, space: source_space, user: user)
+        end
```

Comment:
Looks like this context and before are redundant and can be removed.

Reply:
Sharing a service instance requires Space developer access in both a source and target space. The test setups up the user as a space developer in the source and then table tests their access to the target space. The second context does the opposite. 

So we need both these contexts to have a complete set of permutations.

Reply:
@deniseyu whoops, yep missed that the line farther down was targeting a different space.